### PR TITLE
Reset

### PR DESCRIPTION
**Pull request by timsifive**
Rewrite reset. Now it works on multicore systems, too.

### CODE REVIEW

File: src/target/riscv/riscv-013.c [hunk 1]
```diff
@@ -1187,23 +1185,44 @@ static int examine(struct target *target)
...
+
 	} else {
-		LOG_DEBUG("TARGET RESET HALT NOT SET");
-		control = set_field(control, DMI_DMCONTROL_HALTREQ, 0);
+		// Reset just this hart.
+		uint32_t control = set_field(control_base, DMI_DMCONTROL_HARTSEL,
+				r->current_hartid);
+		control = set_field(control, DMI_DMCONTROL_HALTREQ,
+				target->reset_halt ? 1 : 0);
+		control = set_field(control, DMI_DMCONTROL_HARTRESET, 1);
```
Comment (mwachs5) [inline src/target/riscv/riscv-013.c:1221]: HARTRESET is optional. How do you handle the case when it doesn't work?

**Review by mwachs5 — CHANGES_REQUESTED** [review]
I'm not sure how this works if you have something that doesn't use rtos, is only a single hart, and doesn't have HARTRESET functionality implemented.

File: src/target/riscv/riscv-013.c [hunk 2]
```diff
@@ -1187,23 +1185,44 @@ static int examine(struct target *target)
...
-	control = set_field(control, DMI_DMCONTROL_NDMRESET, 1);
-	if (target->reset_halt) {
-		LOG_DEBUG("TARGET RESET HALT SET, ensuring halt is set during reset.");
-		control = set_field(control, DMI_DMCONTROL_HALTREQ, 1);
+
+	uint32_t control_base = set_field(0, DMI_DMCONTROL_DMACTIVE, 1);
+
+	if (target->rtos) {
+		// There's only one target, and OpenOCD thinks each hart is a thread.
+		// We must reset them all.
```
Comment (mwachs5) [inline src/target/riscv/riscv-013.c:1196]: Maybe should update this comment to say "we must reset them all, but we only halt the ones controlled by this target"

File: src/target/riscv/riscv-013.c (hunk 1 above)
Comment (mwachs5) [reply src/target/riscv/riscv-013.c:1221]: I think you should read back HARTRESET and if it's not actually set, you should fall back to writing NDMRESET (potentially with some debug information).

Also, I think that even if you DID have HARTRESET, the user may want to distinguish between "reset all the things" and "reset a single Hart". Is there an OpenOCD way of clarifying that intention?

File: src/target/riscv/riscv-013.c (hunk 2 above)
Comment (timsifive) [reply src/target/riscv/riscv-013.c:1196]: There is only one target in the RTOS case, and that target controls all the harts.

**Comment by timsifive** [comment]
> I think you should read back HARTRESET and if it's not actually set, you should fall back to writing NDMRESET (potentially with some debug information).

Done.

> Also, I think that even if you DID have HARTRESET, the user may want to distinguish between "reset all the things" and "reset a single Hart". Is there an OpenOCD way of clarifying that intention?

I don't think there is. See http://openocd.org/doc/html/General-Commands.html#resetcommand

**Review by mwachs5 — APPROVED** [review]

**Pull request merged**